File: 2022/12/main.js
const fs = require('fs');
const path = require('path');

const input = fs.readFileSync(path.resolve(__dirname, "./input.txt"), 'utf-8');

let start, end;
const map = input.split('\n').map((line, i) => {
    return line.split('').map((c, j) => {
        switch(c) {
            case 'S': {
                start = [j, i];
                return 1;
            }
            case 'E': {
                end = [j, i];
                return 26;
            }
            default: {
                return c.charCodeAt(0) - 96;
            }
        }
    });
});

const h = map.length;
const w = map[0].length;
const queue = [start];
const steps = { [start]: 0 };
const seen = {};
const prevs = {};

while(queue.length) {
    queue.sort((a, b) => { return steps[a] - steps[b]; });
    // console.log('');
    // queue.forEach((n) => console.log(`${n}: ${steps[n]}`));
    // console.log('');
    const [x1, y1] = queue.shift();
    const n_steps = steps[[x1, y1]];
    if(seen[[x1, y1]]) {
        console.error(`seen ${[x1, y1]} before`);
        break;
    }
    seen[[x1, y1]] = true;
    console.log(`visiting ${[x1, y1]}: ${n_steps}`);
    if(x1 === end[0] && y1 === end[1]) {
        console.log('found the end!');
        break;
    }
    [
        [x1 - 1, y1],
        [x1 + 1, y1],
        [x1, y1 - 1],
        [x1, y1 + 1],
    ].forEach(([x2, y2]) => {
        if(x2 < 0) return;
        if(x2 >= w) return;
        if(y2 < 0) return;
        if(y2 >= h) return;
        if(map[y2][x2] - map[y1][x1] < 2) {
            if(steps[[x2, y2]] === undefined || n_steps + 1 < steps[[x2, y2]]) {
                // console.log(`pushing ${[x2, y2]}: ${n_steps + 1} (prev ${steps[[x2, y2]]})`);
                steps[[x2, y2]] = n_steps + 1;
                prevs[[x2, y2]] = [x1, y1];
                queue.push([x2, y2]);
            }
        }
    });
}
console.log(`part 1: ${steps[end]}`);
const p = [[end, map[end[1]][end[0]]]];
let node = end;
while(prevs[node]) {
    node = prevs[node];
    p.push([node, map[node[1]][node[0]]]);
}
p.reverse();
console.log(p);

File: 2022/12/src/main.rs
use std::collections::HashMap;
use std::collections::VecDeque;
use std::fs;

fn main() {
    let input = fs::read_to_string("input.txt").expect("error reading input file");
    let mut start: Option<(isize, isize)> = None;
    let mut end: Option<(isize, isize)> = None;
    let map: Vec<Vec<isize>> = input
        .lines()
        .enumerate()
        .map(|(i, line)| {
            line.chars()
                .enumerate()
                .map(|(j, char)| match char {
                    'S' => {
                        start = Some(((i as isize), (j as isize)));
                        1
                    }
                    'E' => {
                        end = Some(((i as isize), (j as isize)));
                        26
                    }
                    c => (c as isize) - 96,
                })
                .collect()
        })
        .collect();
    let start: (isize, isize) = start.unwrap();
    let end: (isize, isize) = end.unwrap();
    let h = map.len() as isize;
    let w = map[0].len() as isize;

    {
        let mut queue: VecDeque<(isize, isize)> = VecDeque::new();
        queue.push_back(start);
        let mut steps: HashMap<(isize, isize), usize> = HashMap::from([(start, 0)]);
        loop {
            queue
                .make_contiguous()
                .sort_by(|a, b| steps.get(a).unwrap().cmp(steps.get(b).unwrap()));
            let (i, j) = queue.pop_front().unwrap();
            let n_steps = steps.get(&(i, j)).copied().unwrap();
            // println!("visiting {},{}: {}", j, i, n_steps);
            if (i, j) == end {
                // println!("found the end!");
                println!("part 1: {}", steps.get(&end).unwrap());
                break;
            }
            for (x, y) in [(i - 1, j), (i + 1, j), (i, j - 1), (i, j + 1)] {
                if x < 0 {
                    continue;
                }
                if x >= h {
                    continue;
                }
                if y < 0 {
                    continue;
                }
                if y >= w {
                    continue;
                }
                if (map[x as usize][y as usize]) - (map[i as usize][j as usize]) < 2 {
                    match steps.get(&(x, y)) {
                        Some(n) if n_steps + 1 >= *n => {
                            // println!("skipping {},{} because {} >= {}", y, x, n_steps + 1, n);
                        }
                        _ => {
                            // println!("pushing {},{}: {}", y, x, n_steps + 1);
                            steps.insert((x, y), n_steps + 1);
                            queue.push_back((x, y));
                        }
                    };
                }
            }
        }
    }

    {
        let mut queue: VecDeque<(isize, isize)> = VecDeque::new();
        queue.push_back(end);
        let mut steps: HashMap<(isize, isize), usize> = HashMap::from([(end, 0)]);
        loop {
            queue
                .make_contiguous()
                .sort_by(|a, b| steps.get(a).unwrap().cmp(steps.get(b).unwrap()));
            let (i, j) = queue.pop_front().unwrap();
            let n_steps = steps.get(&(i, j)).copied().unwrap();
            // println!("visiting {},{}: {}", j, i, n_steps);
            if map[i as usize][j as usize] == 1 {
                // println!("found the end!");
                println!("part 2: {}", steps.get(&(i, j)).unwrap());
                break;
            }
            for (x, y) in [(i - 1, j), (i + 1, j), (i, j - 1), (i, j + 1)] {
                if x < 0 {
                    continue;
                }
                if x >= h {
                    continue;
                }
                if y < 0 {
                    continue;
                }
                if y >= w {
                    continue;
                }
                if (map[i as usize][j as usize]) - (map[x as usize][y as usize]) < 2 {
                    match steps.get(&(x, y)) {
                        Some(n) if n_steps + 1 >= *n => {
                            // println!("skipping {},{} because {} >= {}", y, x, n_steps + 1, n);
                        }
                        _ => {
                            // println!("pushing {},{}: {}", y, x, n_steps + 1);
                            steps.insert((x, y), n_steps + 1);
                            queue.push_back((x, y));
                        }
                    };
                }
            }
        }
    }
}
